### PR TITLE
Validation function modification

### LINKED ISSUE #38: Move assertions from constructor functions to validation functions
The below issue applies at the moment to the:
- [x] StanModule Object
- [x] HazardLink Object 

Currently all of our objects assertions are done in the constructor functions i.e.

```
setMethod(
    f = "initialize",
    signature = "StanModule",
    definition = function(
        .Object,
        ...,
    ) {
        assert_that(
            is.character(functions),
            is.character(data),
            is.character(parameters),
            is.character(transformed_parameters),
            is.character(generated_quantities),
            msg = paste(
                "`Functions`, `data`, `parameters`, `transformed_parameters` and",
                "`generated_quantities` must be character vectors"
            )
        )
```

I imagine some of these (though not all of them) should probabally be moved to the S4 validation function. From the [advanced R](https://adv-r.hadley.nz/s4.html) book:

image

### PR DESCRIPTION
**Pull request by charksong**
closes #38

### CODE REVIEW

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:68]: I don't think you need to provide a documentation file for this function

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:56]: from the advanced R book ( [link](https://adv-r.hadley.nz/s4.html#validator) )  I think this function is supposed to return  TRUE (if it is valid) or a string (if it isn't valid). I don't think you are supposed to call the stop function yourself, the validation takes care of that for you behind the scenes

Comment (charksong) [reply R/stan_module.R:56]: Hi Criag, by stop function I thought the codes may be more concise than the book's did
image
if stop function is redundant for some concern, i.e. computational cost, I will follow the book's default guide

File: R/hazard_link.R
Comment (gowerc) [inline R/hazard_link.R:45]: I think these overwrite each other,  I don't think they append for example:

```
myc <- setClass("myc", slots = list(x = "character"))

setValidity("myc", function(object) length(object@x) > 2)
setValidity("myc", function(object) length(object@x) == 1)

x <- myc(x = "a")
```

Doesn't error because the second condition takes precedence

**Comment by gowerc** [comment]
@charksong ,  Lookings like the automated unit tests are failing due to the remove of `assertthat` from the imports. I think we need to list it as an import for some of the other functions for example I think it is used by `is_file()` in `stan_module.R`

File: NAMESPACE
Comment (gowerc) [inline NAMESPACE]: @charksong , Looks like something has gone wrong here. You've included the merge conflict  brackets (the `<<<<<<` and `=====` symbols)  and you also appear to have imported random words from the assertthat library that I'm pretty sure are not functions within it :)

File: R/hazard_link.R
Comment (gowerc) [inline R/hazard_link.R:49]: You appear to have committed the merge conflict boundary flags

File: NAMESPACE
Comment (gowerc) [inline NAMESPACE:49]: @charksong , Looks like something is going wrong with your documentation as it appears to have added a whole sentence to the NAMESPACE file rather than specific functions.  At least I'm pretty sure there is no `will` function coming from the `assertthat` library :)

File: R/stan_module.R
Comment (gowerc) [inline R/stan_module.R:190]: @charksong  I think the issue I mentioned above about the NAMESPACE file is coming from this line. I think this needs to be placed below the function description

Comment (charksong) [inline R/stan_module.R:190]: Hi Craig @gowerc, though the checks passed however I imported random words from the assertthat library, I think the problem is from my moved `#' @importFrom assertthat assert_that` on line 45 before S4 function directly to that before S3 function `is_file`, and it's a matter of roxygen2's syntax. If my guess is right, I'm going to read carefully roxygen2's guide.

**Comment by charksong** [comment]
Thanks a lot Craig! I think I'm going to learn more about roxygen2's syntax.

**Comment by charksong** [comment]
Hi Craig@gowerc , I think this finally may be closed. Thanks for review.

File: R/hazard_link.R
Comment (danielinteractive) [inline R/hazard_link.R:33]: same here

**Review by danielinteractive — COMMENTED** [review]
thanks Marco, please see inside

**Review by danielinteractive — APPROVED** [review]
thanks Marco!

**Comment by danielinteractive** [comment]
@charksong I see that we got errors on the checks, I guess we need a new issue that addresses and fixes those .
please next time don't merge when there are check errors.
@gowerc fyi - maybe it would be better to force check passes and otherwise don't allow merges
image

**Comment by charksong** [comment]
@danielinteractive @gowerc 
My apologies, maybe we could apply `git reflog` and `git reset --hard 6bcedd2e7d7d399e3aaccac52f29ca7651036b44` to back on the last version before this merging by George.

**Comment by danielinteractive** [comment]
@charksong hm I think the merge was done by you, not George - but does not matter. I would not reset for this one, just fix the problem is sufficient

**Comment by gowerc** [comment]
@danielinteractive , I had enabled that option. Not sure how this still went through :S

image

I think from the log it looks like you had approved it ?

**Comment by danielinteractive** [comment]
Yes I approved it because I assume checks will be and need to be fixed

**Comment by danielinteractive** [comment]
@cicdguy do you have any idea? we are out of ideas :-)

**Pull request merged**
